pkg/controller/keycloak/keycloak_controller.go

```diff
@@ -51,27 +54,44 @@ func newReconciler(mgr manager.Manager) reconcile.Reconciler {
 }
 
 // add adds a new Controller to mgr with r as the reconcile.Reconciler
-func add(mgr manager.Manager, r reconcile.Reconciler) error {
+func add(mgr manager.Manager, r reconcile.Reconciler, autodetectChannel chan schema.GroupVersionKind) error {
 	// Create a new controller
-	c, err := controller.New("keycloak-controller", mgr, controller.Options{Reconciler: r})
+	c, err := controller.New(ControllerName, mgr, controller.Options{Reconciler: r})
 	if err != nil {
 		return err
 	}
 
-	// Some resource types may not exist on the cluster when this operator starts.
-	// This autodetect background ticker continually checks if the resource type exists and setup the controller watch if it does
-	if d, err := common.NewAutoDetect(mgr, c); err != nil {
-		log.Error(err, "failed to start the background process to auto-detect the operator capabilities")
-	} else {
-		d.Start()
-	}
-
 	// Watch for changes to primary resource Keycloak
 	err = c.Watch(&source.Kind{Type: &keycloakv1alpha1.Keycloak{}}, &handler.EnqueueRequestForObject{})
 	if err != nil {
 		return err
 	}
 
+	// Setting up a listener for events on the channel from autodetect
+	go func() {
+		for gvk := range autodetectChannel {
+			// Check if this channel event was for the PrometheusRule resource type
+			if gvk.String() == monitoringv1.SchemeGroupVersion.WithKind(monitoringv1.PrometheusRuleKind).String() {
+				watchSecondaryResource(c, gvk, &monitoringv1.PrometheusRule{}) // nolint
+			}
+
+			// Check if this channel event was for the ServiceMonitor resource type
+			if gvk.String() == monitoringv1.SchemeGroupVersion.WithKind(monitoringv1.ServiceMonitorsKind).String() {
+				watchSecondaryResource(c, gvk, &monitoringv1.ServiceMonitor{}) // nolint
+			}
+
+			// Check if this channel event was for the GrafanaDashboard resource type
+			if gvk.String() == integreatlyv1alpha1.SchemeGroupVersion.WithKind(integreatlyv1alpha1.GrafanaDashboardKind).String() {
+				watchSecondaryResource(c, gvk, &integreatlyv1alpha1.GrafanaDashboard{}) // nolint
+			}
+
+			// Check if this channel event was for the Route resource type
+			if gvk.String() == routev1.SchemeGroupVersion.WithKind(common.RouteKind).String() {
+				_ = watchSecondaryResource(c, gvk, &routev1.Route{})
```

Comment:
This PR currently breaks the Openshift detection because we just check there if a watch for a Route kind exists (in cluster_state.go). But the name of the key in the state manager is not tied to the controller name. What about adding an extra key value pair `routeDetected: true` to the state manager if a route was detected?

Reply:
Good catch @pb82 . Pushed another commit now